tucan/io/molfile_reader.py

```diff
@@ -183,24 +192,59 @@ def _parse_bond_block_molfile3000(lines: list[list[str]]) -> list[tuple[int, int
             f'Expected "END BOND" on line {bond_block_offset + bond_count + 1}, found "{end_bond_str}"'
         )
 
-    bonds = {
-        (
-            # atom-indices are zero-based
-            int(line[4]) - 1,
-            int(line[5]) - 1,
-        ): _parse_bond_props(line)
-        for line in lines[bond_block_offset : bond_block_offset + bond_count]
-    }
+    bonds = {}
+    for line in lines[bond_block_offset : bond_block_offset + bond_count]:
+        atom1_index = int(line[4]) - 1
+        atom2_index = int(line[5]) - 1
+        bond_props = _parse_bond_props(line)
+
+        atom1_is_star = atom1_index in star_atoms
+        atom2_is_star = atom2_index in star_atoms
+        if atom1_is_star and atom2_is_star:
+            raise MolfileParserException(
+                f'Two "star" atoms (index {atom1_index + 1} and {atom2_index + 1}) may not be connected'
+            )
+        elif atom1_is_star:
+            bond_tuples = _parse_bond_line_with_star_atom(line, atom2_index)
+        elif atom2_is_star:
+            bond_tuples = _parse_bond_line_with_star_atom(line, atom1_index)
+        else:
+            bond_tuples = [(atom1_index, atom2_index)]
+
+        for t in bond_tuples:
+            bonds[t] = bond_props.copy()
 
     return bonds
 
 
 def _parse_bond_props(line: list[str]) -> dict:
-    return {"bond_type": line[3]}
+    return {"bond_type": int(line[3])}
+
+
+def _parse_bond_line_with_star_atom(
+    line: list[str], start_atom_index: int
+) -> list[tuple[int, int]]:
+    for index, token in enumerate(line):
+        if token.startswith("ENDPTS=("):
+            endpts_index = index
+            number_of_endpts = int(token.split("=(")[1])
+            break
+    # silently ignore everything that has no ENDPTS (e.g. use of star atoms in polymers)
+    else:
+        return []
+
+    endpoints = [
+        int(token) for token in line[endpts_index + 1 : endpts_index + number_of_endpts]
+    ]
+    if not line[endpts_index + number_of_endpts].endswith(")"):
+        raise MolfileParserException("Expected end of ENDPTS block")
+    endpoints.append(int(line[endpts_index + number_of_endpts][:-1]))
+
+    return [(start_atom_index, end_atom_index - 1) for end_atom_index in endpoints]
```

Comment:
Could we use a regex instead?

```Python
def _parse_bond_line_with_star_atom(
    line: list[str], start_atom_index: int
) -> list[tuple[int, int]]:

    endpts_pattern = re.compile(r"ENDPTS=\(.+\)")
    endpts_match = endpts_pattern.search(" ".join(line))
    if endpts_match is None:
        # silently ignore everything that has no ENDPTS (e.g. use of star atoms in polymers)
        return []
    endpts_token = endpts_match.group()
    endpoints = [int(endpts_index) for endpts_index in endpts_token[9:-1].split()]
    return [(start_atom_index, end_atom_index - 1) for end_atom_index in endpoints]
```

With the regex, there'd be no need for this either (the corresponding test case in `test_molfile_with_star_atoms_raises_exception` would be obsolete):
```Python
if not line[endpts_index + number_of_endpts].endswith(")"):
    raise MolfileParserException("Expected end of ENDPTS block")
```